protocol/contracts/GenericZap.sol

```diff
@@ -0,0 +1,918 @@
+  event TokenRecovered(address token, address to, uint256 amount);
+  event ZappedLPCurve(address indexed recipient, address fromToken, uint256 liquidity, uint256[] amounts);
+  event ZappedLiquidityBalancerPool(address indexed recipient, address fromToken, uint256 fromAmount, uint256[] maxAmountsIn);
+
+  constructor (
+    address _router
+  ) {
+    uniswapV2Router = IUniswapV2Router(_router);
+  }
+
+  /**
+   * @notice recover token or ETH
+   * @param _token token to recover
+   * @param _to receiver of recovered token
+   * @param _amount amount to recover
+   */
+  function recoverToken(address _token, address _to, uint256 _amount) external onlyOwner {
+    require(_to != address(0), "Invalid receiver");
+    if (_token == address(0)) {
+      // this is effectively how OpenZeppelin transfers eth
+      require(address(this).balance >= _amount, "Address: insufficient balance");
+      (bool success,) = _to.call{value: _amount}(""); 
+      require(success, "Address: unable to send value");
+    } else {
+      _transferToken(IERC20(_token), _to, _amount);
+    }
+    
+    emit TokenRecovered(_token, _to, _amount);
+  }
+
+  function setUniswapV2Router(address _router) external onlyOwner {
+    uniswapV2Router = IUniswapV2Router(_router);
+    emit UniswapV2RouterSet(_router);
+  }
+
+  /**
+   * @notice This function zaps ETH and ERC20 tokens to Temple token
+   * @param fromToken The token used for entry (address(0) if ether)
+   * @param fromAmount The amount of fromToken to zap
+   * @param toToken Exit token
+   * @param amountOutMin The minimum acceptable quantity of TEMPLE to receive
+   * @param swapTarget Execution target for the swap
+   * @param swapData DEX data
+   * @return amountOut Amount of tokens received
+   */
+  function zapIn(
+    address fromToken,
+    uint256 fromAmount,
+    address toToken,
+    uint256 amountOutMin,
+    address swapTarget,
+    bytes calldata swapData
+  ) external payable whenNotPaused returns (uint256 amountOut) {
+    require(approvedTargets[fromToken][swapTarget] == true, "Unsupported token/target");
+
+    amountOut = _zapIn(
+      fromToken,
+      fromAmount,
+      toToken,
+      amountOutMin,
+      swapTarget,
+      swapData
+    );
+  }
+
+  function zapLiquidityBalancerPoolFor(
+    address _fromToken,
+    uint256 _fromAmount,
+    bytes32 _poolId,
+    address _recipient,
+    address _swapTarget,
+    bytes memory _swapData,
+    ZapLiquidityRequest memory _zapLiqRequest,
+    IBalancerVault.JoinPoolRequest memory _request
+  ) external payable whenNotPaused {
+    require(approvedTargets[_fromToken][_swapTarget] == true, "Unsupported token/target");
+
+    _pullTokens(_fromToken, _fromAmount);
+
+    _zapLiquidityBalancerPool(
+      _fromToken,
+      _fromAmount,
+      _poolId,
+      _recipient,
+      _swapTarget,
+      _swapData,
+      _zapLiqRequest,
+      _request
+    );
+  }
+
+  function zapLiquidityBalancerPool(
```

Comment:
not quite sure i understand what you mean in the first part. there's already a `zapLiquidityBalancerPoolFor` function (including the other lp pools).

I will add a `zapInFor`

Reply:
I guess typically in the non-contract space I'd write as:
```
  function zapLiquidityBalancerPoolFor(...) public payable whenNotPaused {
    require(approvedTargets[_fromToken][_swapTarget] == true, "Unsupported token/target");

    _pullTokens(_fromToken, _fromAmount);

    _zapLiquidityBalancerPool(
      _fromToken,
      _fromAmount,
      _poolId,
      _recipient,
      _swapTarget,
      _swapData,
      _zapLiqRequest,
      _request
    );
  }

  function zapLiquidityBalancerPool(...) external payable {
    zapLiquidityBalancerPoolFor(..., msg.sender, ...);
  }
```

So the requires, modifiers, pullTokens etc are guaranteed to be consistent / non-duplicated logic too. I understand there's a tiny bit more gas because another function redirect, but would expect that to be very small.